csharp/src/ParallelSelectExt.cs

```diff
@@ -0,0 +1,368 @@
+
+      // Async closure that waits for the task x and releases the semaphore
+      async Task<T> TaskLambda()
+      {
+        var res = await x.ConfigureAwait(false);
+        sem.Release();
+        return res;
+      }
+
+      // We can enqueue the task
+      queue.Enqueue(Task.Run(TaskLambda,
+                             cancellationToken));
+
+      next = semAcquire;
+    }
+
+    // We finished iterating over the inputs and
+    // must now wait for all the tasks in the queue
+    foreach (var task in queue)
+    {
+      // Dequeue a new task and wait for its result
+      // As the task is wrapped in a Task.run with the cancellation token,
+      // It will be cancelled as soon cancellation is triggered
+      using (task)
+      {
+        yield return await task.ConfigureAwait(false);
+      }
+    }
+  }
+
+  /// <summary>
+  ///   Iterates over the input enumerable and spawn multiple parallel tasks that call `func`.
+  ///   The maximum number of tasks in flight at any given moment is given in the `parallelTaskOptions`.
+  ///   All results are collected in-order.
+  /// </summary>
+  /// <param name="enumerable">Enumerable to iterate on</param>
+  /// <param name="parallelTaskOptions">Options (eg: parallelismLimit, cancellationToken)</param>
+  /// <param name="func">Function to spawn on the enumerable input</param>
+  /// <typeparam name="TInput">Type of the inputs</typeparam>
+  /// <typeparam name="TOutput">Type of the outputs</typeparam>
+  /// <returns>Asynchronous results of func over the inputs</returns>
+  [PublicAPI]
+  public static IAsyncEnumerable<TOutput> ParallelSelect<TInput, TOutput>(this IEnumerable<TInput>    enumerable,
+                                                                          ParallelTaskOptions         parallelTaskOptions,
+                                                                          Func<TInput, Task<TOutput>> func)
+    => ParallelWait(enumerable.Select(func),
+                    parallelTaskOptions);
+
+  /// <summary>
+  ///   Iterates over the input enumerable and spawn multiple parallel tasks that call `func`.
+  ///   The maximum number of tasks in flight at any given moment is given in the `parallelTaskOptions`.
+  ///   All results are collected in-order.
+  /// </summary>
+  /// <param name="enumerable">Enumerable to iterate on</param>
+  /// <param name="parallelTaskOptions">Options (eg: parallelismLimit, cancellationToken)</param>
+  /// <param name="func">Function to spawn on the enumerable input</param>
+  /// <typeparam name="TInput">Type of the inputs</typeparam>
+  /// <typeparam name="TOutput">Type of the outputs</typeparam>
+  /// <returns>Asynchronous results of func over the inputs</returns>
+  [PublicAPI]
+  public static IAsyncEnumerable<TOutput> ParallelSelect<TInput, TOutput>(this IAsyncEnumerable<TInput> enumerable,
+                                                                          ParallelTaskOptions           parallelTaskOptions,
```

Comment:
This is th optional argument. It should be last with proper default value. This allows to remove another implementation.

Reply:
From what I saw, the convention is more to use the functor as last argument rather than putting the "default parameter" at the end: https://learn.microsoft.com/en-us/dotnet/api/system.threading.tasks.parallel.foreach

Reply:
I'd put the optional parameter at the end